### PR TITLE
fix prune method

### PR DESCRIPTION
**Pull request by giolaza**

### CODE REVIEW

**Review by willpower232 — CHANGES_REQUESTED** [review]
please see progress in #948

**Comment by giolaza** [comment]
> please see progress in #948

#948  is better because it more dynamic

File: src/Drivers/Database.php [hunk 1]
```diff
@@ -23,10 +23,26 @@ public function audit(Auditable $model): ?Audit
...
+            $class = get_class($model->audits()->getModel());
+            $auditKeyName = (new $class)->getKeyName();
+
+            $forRemoval = $model->audits()
                 ->latest()
-                ->offset($threshold)->limit(PHP_INT_MAX)
-                ->delete() > 0;
+                ->get()
+                ->slice($threshold)
+                ->pluck($auditKeyName);
```
Comment (parallels999) [inline src/Drivers/Database.php:33]: It seems to me that it is more efficient to put the `pluck` above the `slice`, replacing the `get`

File: src/Drivers/Database.php [hunk 2]
```diff
@@ -23,10 +23,26 @@ public function audit(Auditable $model): ?Audit
...
+
+            if (!$forRemoval->isEmpty()) {
+                $forRemovalChunks = $forRemoval->chunk(Config::get('audit.placeholders_limit', 50000));
+                $answer = false;
+                foreach ($forRemovalChunks as $chunk) {
+                    $answer = $answer || $model->audits()
```
Comment (parallels999) [inline src/Drivers/Database.php:39]: Maybe I don't understand this,
after the first `chunk`, `answer` would be `true`,
then in the second chunk(and later), the delete would no longer be executed

**Review by parallels999 — COMMENTED** [review]
You also need to add tests to corroborate that the chuncks are working.

File: src/Drivers/Database.php (hunk 2 above)
Comment (parallels999) [reply src/Drivers/Database.php]: Also, if the first chunk is executed, and the second for some reason is not, would it return `true`/`false` even though only a portion of the records were deleted?

File: src/Drivers/Database.php [hunk 3]
```diff
@@ -23,10 +23,26 @@ public function audit(Auditable $model): ?Audit
...
+            $forRemoval = $model->audits()
                 ->latest()
-                ->offset($threshold)->limit(PHP_INT_MAX)
-                ->delete() > 0;
+                ->get()
+                ->slice($threshold)
+                ->pluck($auditKeyName);
+
+            if (!$forRemoval->isEmpty()) {
+                $forRemovalChunks = $forRemoval->chunk(Config::get('audit.placeholders_limit', 50000));
```
Comment (parallels999) [inline src/Drivers/Database.php:36]: If you have thousands of records, they will all be stored in memory as models, wouldn't that be less efficient than a `leftJoin`?

File: src/Drivers/Database.php (hunk 2 above)
Comment (giolaza) [reply src/Drivers/Database.php]: any ideas? I thought if some deletion was  performed in any case return true

File: src/Drivers/Database.php [hunk 4]
```diff
@@ -23,10 +23,26 @@ public function audit(Auditable $model): ?Audit
...
         if (($threshold = $model->getAuditThreshold()) > 0) {
-            return $model->audits()
+            $class = get_class($model->audits()->getModel());
+            $auditKeyName = (new $class)->getKeyName();
+
+            $forRemoval = $model->audits()
                 ->latest()
-                ->offset($threshold)->limit(PHP_INT_MAX)
-                ->delete() > 0;
+                ->skip($threshold)
```
Comment (angeljqv) [inline src/Drivers/Database.php:31]: skip and take are alias of offset and limit

**Comment by giolaza** [comment]
think this topic now is closed

**Pull request closed without merging**